File: variacio/varism.cs
using System;
using System.Diagnostics;

namespace Variaciok
{
    class VarIsm
    {
        public static void Main(string[] args)
        {
            Console.WriteLine(DateTime.Now.ToString());
            int n = 120, k = 5;
            //int n = 4, k = 3;
            int[] arr = new int[n];
            for (int i = 0; i < n; ++i) arr[i] = i + 1;
            Stopwatch sw = new Stopwatch();
            long t_1 = 0, t_2 = 0, t_3 = 0;
            GC.Collect();
            sw.Start();
            IsmVariacio(arr, k); //kb. 35 sec.
            t_1 = sw.ElapsedMilliseconds;
            GC.Collect();
            t_2 = sw.ElapsedMilliseconds;
            IsmVar(arr, k); //kb. 42 sec.
            t_3 = sw.ElapsedMilliseconds;
            Console.WriteLine("Eltelt idő: {0}", t_1);
            Console.WriteLine("Eltelt idő: {0}", t_3 - t_2);
        }

        static void IsmVariacio(int[] arr, int k)
        {
            int n = arr.Length;
            int[] tomb = new int[k];
            for (int i = 0; i < k; ++i) tomb[i] = 0;
            int j = k - 1;
            do
            {

                /*for (int i = 0; i < k; ++i) Console.Write("{0} ", arr[tomb[i]]);
                Console.WriteLine("");*/

                ++tomb[j];
                if (tomb[j] > n - 1)
                {
                    while (j > -1)
                    {
                        if (tomb[j] >= n - 1)
                        {
                            --j;
                            if (j == -1) return;
                        }
                        else break;
                    }
                    ++tomb[j];
                    ++j;
                    while (j < k)
                    {
                        tomb[j] = 0;
                        ++j;
                    }
                    --j;
                }
            } while (true);
        }

        static void IsmVar(int[] arr, int k)
        {
            int n = arr.Length;
            int[] tomb = new int[k + 1];
            int i;
            for (i = 1; i <= k; i++) tomb[i] = 1;
            tomb[k] = 0;
            i = k;
            while (true)
            {
                if (tomb[i] == n)
                {
                    i--;
                    if (i == 0) return;
                }
                else
                {
                    tomb[i]++;
                    while (i < k)
                    {
                        i++;
                        tomb[i] = 1;
                    }

                    /*for (int j = 1; j <= k; j++) Console.Write("{0} ", arr[tomb[j] - 1]);
                    Console.WriteLine("");*/
                }
            }
        }
    }
}

File: variacio/varism.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/timeb.h>

long long db = 0;

void IsmVariacio(int arr[], int n, int k)
{
    int* tomb = (int*)calloc(k, sizeof(int));
    int i;
    int j = k - 1;
    do
    {
        ++db;
        /*for (int i = 0; i < k; ++i) printf("%d ", arr[tomb[i]]);
        printf("\n");*/

        ++tomb[j];
        if (tomb[j] > n - 1)
        {
            while (j > -1)
            {
                if (tomb[j] >= n - 1)
                {
                    --j;
                    if (j == -1) return;
                }
                else break;
            }
            ++tomb[j];
            ++j;
            while (j < k)
            {
                tomb[j] = 0;
                ++j;
            }
            --j;
        }
    } while (1);
    free(tomb);
}

void IsmVar(int arr[], int n, int k)
{
    int tomb[50], i;
    for (i = 1; i <= k; i++) tomb[i] = 1;
    tomb[k] = 0;
    i = k;
    while (1) {
        if (tomb[i] == n) {
            i--;
            if (i == 0) break;
        }
        else {
            tomb[i]++;
            while (i < k) {
                i++;
                tomb[i] = 1;
            }

            ++db;
            /*int j;
            for (j = 1; j <= k; j++) printf("%d ", arr[tomb[j] - 1]);
            printf("\n");*/
        }
    }
}


long timediff(struct timeb* start, struct timeb* end)
{
    long seconds;
    seconds = (long)(end->time - start->time);
    start->millitm = end->millitm - start->millitm;
    if (0 > start->millitm) {
        start->millitm += 1000;
        seconds--;
    }
    return seconds;
}

int main()
{
    struct timeb start, end;
    long seconds, seconds2;
    int militm, militm2;
    int arr[500];
    int i, n = 120, k = 5;
    for (i = 0; i < n; i++) arr[i] = i + 1;
    ftime(&start);
    IsmVariacio(arr, n, k); //kb 36 sec.
    ftime(&end);
    seconds = timediff(&start, &end);
    militm = start.millitm;
    db = 0;
    ftime(&start);
    IsmVar(arr, n, k); //kb. 30 sec.
    ftime(&end);
    seconds2 = timediff(&start, &end);
    militm2 = start.millitm;
    printf("Eltelt ido: %ld.%03d masodperc\n", seconds, militm);
    printf("Eltelt ido: %ld.%03d masodperc\n", seconds2, militm2);
    return 0;
}
